src/ImageSharp/Processing/Processors/Binarization/AdaptiveThresholdProcessor{TPixel}.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Buffers;
+using System.Runtime.CompilerServices;
+
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Binarization
+{
+    /// <summary>
+    /// Performs Bradley Adaptive Threshold filter against an image.
+    /// </summary>
+    internal class AdaptiveThresholdProcessor<TPixel> : ImageProcessor<TPixel>
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        private readonly AdaptiveThresholdProcessor definition;
+        private readonly PixelOperations<TPixel> pixelOpInstance;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveThresholdProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="configuration">The configuration which allows altering default behaviour or extending the library.</param>
+        /// <param name="definition">The <see cref="AdaptiveThresholdProcessor"/> defining the processor parameters.</param>
+        /// <param name="source">The source <see cref="Image{TPixel}"/> for the current processor instance.</param>
+        /// <param name="sourceRectangle">The source area to process for the current processor instance.</param>
+        public AdaptiveThresholdProcessor(Configuration configuration, AdaptiveThresholdProcessor definition, Image<TPixel> source, Rectangle sourceRectangle)
+            : base(configuration, source, sourceRectangle)
+        {
+            this.pixelOpInstance = PixelOperations<TPixel>.Instance;
+            this.definition = definition;
+        }
+
+        /// <inheritdoc/>
+        protected override void OnFrameApply(ImageFrame<TPixel> source)
+        {
+            var intersect = Rectangle.Intersect(this.SourceRectangle, source.Bounds());
+
+            Configuration configuration = this.Configuration;
+            TPixel upper = this.definition.Upper.ToPixel<TPixel>();
+            TPixel lower = this.definition.Lower.ToPixel<TPixel>();
+            float thresholdLimit = this.definition.ThresholdLimit;
+
+            // Used ushort because the values should never exceed max ushort value.
+            ushort startY = (ushort)intersect.Y;
```

Comment:
This doesn't make sense to me. Our working rectangle can definitely exceed that.

Reply:
i have changed it to int